frontend/src/pages/new_question/utils/queryBuilderUtils.js

```diff
@@ -0,0 +1,199 @@
+    index += 1;
+  }
+  return `e${index}`;
+}
+
+function getNumAttachedEdges(q_graph) {
+  const edgeNums = _.transform(q_graph.edges,
+    (result, value) => {
+      result[value.object] = result[value.object] ? result[value.object] + 1 : 1;
+      result[value.subject] = result[value.subject] ? result[value.subject] + 1 : 1;
+    });
+  return edgeNums;
+}
+
+/**
+ * Parse an incoming query graph and find the "root" node
+ *
+ * Rules are:
+ * 1. sort by unpinned and then pinned nodes
+ * 2. sort by total number of edge connections
+ * 3. pick first node in insertion order
+ * @param {object} q_graph valid query graph with nodes and edges
+ * @returns {string} the id of the root node in the query graph
+ */
+function findRootNode(q_graph) {
+  // create nodes object with pinned boolean property
+  const nodes = Object.entries(q_graph.nodes).map(([key, node]) => (
+    {
+      key,
+      pinned: node.id && Array.isArray(node.id) && node.id.length > 0,
+    }
+  ));
+  // create node map of total edge connections
+  const edgeNums = getNumAttachedEdges(q_graph);
+  // split nodes into pinned and unpinned arrays
+  const unpinnedNodes = nodes.filter((node) => !node.pinned && node.key in edgeNums);
+  const pinnedNodes = nodes.filter((node) => node.pinned && node.key in edgeNums);
+  // sort nodes by edge connections, then return the first one in the list
+  // is also first one inserted
+  let root = null;
+  if (unpinnedNodes.length) {
+    unpinnedNodes.sort((a, b) => edgeNums[b.key] - edgeNums[a.key]);
+    root = unpinnedNodes[0].key;
+  } else if (pinnedNodes.length) {
+    pinnedNodes.sort((a, b) => edgeNums[b.key] - edgeNums[a.key]);
+    root = pinnedNodes[0].key;
+  }
+  return root;
+}
+
+/**
+ * Remove all detached sections of graph from root node
+ * @param {object} q_graph deep copy of the query_graph, modifies and returns
+ * @param {string} rootNode root node of graph
+ */
+function trimDetached(q_graph, rootNode) {
+  // all edges start out as disconnected and we'll remove them when we find a connection to root
+  const disconnectedEdges = new Set(Object.keys(q_graph.edges));
+  // we add to connected nodes, starting default with the root node
+  const connectedNodes = new Set([rootNode]);
+  let foundConnections = true;
+  // traverse graph and delete all edges not connected
+  while (foundConnections) {
+    let foundConnection = false;
+    // loop over all connected nodes to find edges that are connected to them
+    [...connectedNodes].forEach((nodeId) => {
+      [...disconnectedEdges].forEach((e) => {
+        const edge = q_graph.edges[e];
+        // if edge is connected to node
+        if (edge.subject === nodeId || edge.object === nodeId) {
+          connectedNodes.add(edge.subject);
+          connectedNodes.add(edge.object);
+          disconnectedEdges.delete(e);
+          foundConnection = true;
+        }
+      });
+    });
+    // no more connected edges, break out
+    if (!foundConnection) {
```

Comment:
I think we should be able to do this without both `foundConnection` and `foundConnections`. If we just merge them and remove this if block, I think it should work?

Reply:
Updated in 82340751311f45bc61d362c3cdaa80f69255de01